SCClassLibrary/JITLib/ProxySpace/NodeProxy.sc

```diff
@@ -87,7 +87,8 @@ NodeProxy : BusPlug {
 	}
 
 	fadeTime {
-		^nodeMap.at(\fadeTime) ? 0.02;
+		if(nodeMap.at(\fadetime).isNil) { nodeMap.put(\fadeTime, 0.02) };
```

Comment:
sorry, i know this is the product of discussion, but i'm not really following the logic here -- why a state change inside of a getter?

Reply:
Thanks for the comment @snappizz – you are right, this is a stupid place to put it. I want to make sure fadeTime is always set correctly in the nodeMap. As it is now, it will only explicitly be in the nodeMap after copy which is what I wanted to fix here. Do you have any ideas on this?

Reply:
There are indeed cases where a getter changes a state, namely when the state is cached (like in a Thunk). But in this case it is definitely not a good thing, because it erases the distinction between an unspecified and a specified value.

Reply:
It would be OK to have a `classvar <>defaultFadeTime`. Then you could test and only copy if `fadeTime != defaultFadeTime`.

Reply:
and then all the places where we have `? 0.02` should become `? defaultFadeTime`.

Reply:
why not automatically set `fadeTime` to 0.02 in the nodeMap when it's initialized? the implicit vs explicit distinction seems like it's adding unnecessary complexity here.

Reply:
good idea, yes, this would simplify the whole thing.

Reply:
The changes would be:

```supercollider
init {
		nodeMap = ProxyNodeMap.new;
		nodeMap.put(\fadeTime, 0.02);
		objects = Order.new;
		loaded = false;
		reshaping = defaultReshaping;
		this.linkNodeMap;
	}

// and:


	fadeTime_ { | dur |
		this.set(\fadeTime, dur ? 0.02)
	}

	

```

we should probably leave this:

```supercollider
fadeTime {
		^nodeMap.at(\fadeTime) ? 0.02
	}
```

because someone could remove the \fadeTime from the nodeMap directly.

Reply:
that's better, although i would recommend avoiding a duplicate magic number 0.02 and instead put that in `classvar <defaultFadeTime`.

the purpose of the `? 0.02`'s on the getter and setter aren't clear to me. why would someone remove the fadetime from nodemap? that sounds like a user error.

also, it seems wrong to me that `.fadeTime = nil` sets the fade time to 0.02. i don't think there's any point to a default argument in a setter.

Reply:
a classvar makes sense, yes.

Thinking again, the `? 0.02` in the *setter* is unnecessary, but not in the *getter*. So far the semantics has been that setting it to nil is resetting it to default. Anything else will break code.

Reply:
let's back up here and do a mini-specification, since i'm getting confused. what is the desired behavior for `.fadeTime` and `.fadeTime_`, and how should `.copy` affect `.nodeMap[\fadeTime]`?

Reply:
sorry, please disregard my now-deleted comment, i misunderstood. if the specification is that `.fadeTime_ = nil` resets it to default, then yes we should retain that behavior.

there's a double negative in your comment -- my assessment is that we need `? 0.02` in the *setter* (to retain the above described behavior), but not in the *getter*. right?

Reply:
Sorry for the double negation. We actually need it in both places.
1. in the setter, because it calls `this.set` which sends the value to the synth, and `nil` would become `0` (unfortunately, there is no way currently to recapture the original default value of a synth either).
2. in the getter, just to be sure, because someone may by using the `NodeMap` directly and setting fadeTime to nil. You are right that this would be a user error, and there are precautions if you replace the NodeMap as a whole (`this.nodeMap_()`), but if you just access the thing and set it there, it has worked so far. I'd rather no touch it.

With regards to the current pull request: it should be fine if the NodeMap is initialized as suggested. I agree that it is cleaner using a `classvar <defaultFadeTime`.

@madskjeldgaard does all this make sense to you?

Reply:
Hi ! Sorry, been away on vacation and now getting back to this. I think it makes sense. I'm going dive into it again now and hopefully finish it :)

thanks for the ideas!

Reply:
Okay see latest commit for changes. Hope I understood you both correctly here, if not, let me know. 

It now passes the tests I had set up as well and the node maps are equal.